### PR TITLE
Example notebook for proton radiography custom source capability

### PR DESCRIPTION
**Pull request by pheuer**
This PR adds an example notebook demonstrating how users can use `SyntheticProtonRadiograph` to create synthetic proton radiographs using custom lists of initialized particles, and how this capability allows the creation of synthetic radiographs with arbitrary source fluence profiles. 

(Previously `load_particles` wasn't used in any of the example notebooks)

This PR also now includes minor changes to the other PRad notebooks too (mostly adding links). 

TODO: 

- [x] Figure out the reason the links aren't working
- [x] Once links are working, add them to the other two PRad notebooks too
- [x] Create explicit tests for `rot_a_to_b()` in mathematics. Requires making a `test_mathematics.py` file.

### CODE REVIEW

**Comment by pheuer** [comment]
@rocco8773 Can you take a look at this one?

**Comment by rocco8773** [comment]
@pheuer Yes, I'll take a look over it.  Give me today/tomorrow to get to it. Thanks.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:7]: Can you shorten the title by at least 1 word?  Currently the title is so long that it's extending beyond the boundaries of the thumbnail.

image

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:14]:
```suggestion
    "[SyntheticProtronRadiography]: ../../api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.rst#syntheticprotonradiograph\n",
    "\n",
    "In the real world the particle source for charged particle radiography experiments is finite in size and often has a non-uniform angular velocity distribution. For this reason, the [SyntheticProtronRadiography] particle tracing algorithm allows users to specify their own initial particle positions and velocities. This example will demonstrate how to use this functionality."
```

Sphinx directives do not work in notebooks, so when referencing code documentation we have to used mardown's linking mechanisms.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:71]: This notebook does not have any mention on how the xyz coords are defined, so we don't have reference for `phi` and `theta`.  I'd also specifically state what `phi` and `theta` correspond to since they are often flipped between different research fields.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:501]: Delete empty cell at the end of the notebook.  This is often missed because Jupyter notebook automatically adds the cell when you execute the last cell in the notebook.  It's just something we have to be aware of when committing new commits.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:432]: For some reason this thumbnail tagging is not working.  I'll continue looking into it, but two pairs will be quicker than one.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:189]:
```suggestion
    "ax.axvline(x=0.95 * v_cent, label=\"-5%\", color=\"C2\")\n",
```

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:216]: If you're going to reference the `_rot_a_to_b` functionality, then it should not be private.   Either remove the reference from the notebook or refactor it to `rot_a_to_b` and properly document.

**Review by rocco8773 — CHANGES_REQUESTED** [review]
@pheuer Here's a partial review of the notebook.  The notebook looks to be in good condition, but there are a few things to address.

* As you can see in my first cell comment, Sphinx's normal linking mechanisms don't work in notebooks.  Instead we have to manually create links using markdown's syntax.  Can you go through your notebook and add these links to documentation whenever you're referencing `plasmapy`'s functionality or any external pkgs like `numpy`.
* This notebook is long enough that it might warrant making a table of contents at the top of the notebook.  If you feel you'd like to go this way, take a look at the [Fit Function notebook](https://docs.plasmapy.org/en/latest/notebooks/analysis/fit_functions.html) and/or [Floating Potential notebook](https://docs.plasmapy.org/en/latest/notebooks/analysis/swept_langmuir/find_floating_potential.html).
* Can you please optimize the plot sizes and font sizes?  More so the font sizes.  The font sizes are a little too small for readability in a web browser.  For tips you can again refer to the Fit Function and Floating Potential notebooks.
* [optional] When plotting the particle distributions, it might be easier for users to interpret the plots if they are normalized to the total number of particles in the distribution `N / No / dSigma`.  If you go this route, `No` should be included somewhere in the plot.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:7]: Sure, let's try just "Synthetic Radiographs with Custom Source Profiles"

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:14]: I find this pretty confusing, can you elaborate? There is no `/api` folder that I can see, nor a `SyntheticProtonRadiograph.rst`, so I assume these are files that are being created when the documentation is being compiled? How do I reference a method like `create_particles`?

Is there a way to get to this reference path from the documentation URL? They look similar...
```
https://docs.plasmapy.org/en/latest/api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.html#plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph
```

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:216]: I definitely agree, but it's such a general function that I don't think it belongs in `proton_radiography.py`. Where should we put it?

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:71]: I'll specify earlier that the source-detector axis is parallel to Z, and I'll explicitly define theta and phi as the polar and azimuthal angles respectively. I could consider a diagram, but that might be overkill?

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:444]: Hm, no idea, I checked and this `metadata` is identical to what I have in the main proton radiography notebook that seems to be working.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: Everything in the `docs/api/` folder is auto generated by the `automodsumm` and `autosummary` directives.  That's why you don't see the files until you do a build of the documentation.  The naming scheme of the `api/` files is the same as what we use for the `api_static/` files, `plasmapy.mod.mod.Class`, `plasmapy.mod.mod.func`, etc.  The `#` at the end tells the build you're trying to reference a specific heading in that file.  To reference the whole page then you just exclude the `#` part.

image

---
Here's the [`nbsphinx` documentation on this topic.](https://nbsphinx.readthedocs.io/en/0.8.2/markdown-cells.html#Links-to-Domain-Objects)

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: I'll leave that decision to you, but when it comes to defining coordinates it's always easier to understand the coord system via a diagram opposed to words.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:444]: I did the same and couldn't see why.  I'll keep investigating.  I'm sure we'll find the culprit as we get through the review.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:228]: 🤔 It could possibly be put in `plasmapy.formulary.mathematics`.  I think that would make sense.

Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: Sounds good to me!

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (StanczakDominik) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:494]: Ha, third pair's the charm! At first I thought this was a question of [multiple outputs from a cell](https://nbsphinx.readthedocs.io/en/0.8.2/gallery/multiple-outputs.html), and it kind of is. tqdm is the issue - "stream" output seems to be the on-the-fly progress bar animation. Every text line of the ASCII progress bar seems to be recorded as an output.

To sidestep this, I just split this cell into one for the simulation and one for the plot.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (StanczakDominik) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: I'll have to quote that comment into the doc guide...

Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: I've tried to do this, but I must still be doing something wrong? I have

```
[SyntheticProtronRadiography]: ../../api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.rst#plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph

[SyntheticProtronRadiography] 
```

but CI raises this error

```
Warning, treated as error:
/home/runner/work/PlasmaPy/PlasmaPy/docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:20:File not found: '../api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.rst#plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph'
ERROR: InvocationError for command /home/runner/work/PlasmaPy/PlasmaPy/.tox/build_docs/bin/sphinx-build docs docs/_build/html -W -b html (exited with code 2)
```

**Comment by pheuer** [comment]
Note that I've made a couple minor edits to another tutorial notebook (`proton_radiography_particle_tracing`) here too:
1) Shortened the title so it fits in the box below the thumbnail
2) Fixed a warning caused by the newest Matplotlib version

I'd like to add nice documentation links to that notebook too, once we figure out how to make them work on the source profile notebook...

**Comment by StanczakDominik** [comment]
@pheuer just noting - this looks on track for 0.6 :) is there anything off with the thumbnails, still? That last commit message looks concerning.

**Comment by StanczakDominik** [comment]
Paging @rocco8773 the link king ;)

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb]: Your link is going back one directory too far, `../../` is sending you back to the repor root directory instead of `/docs/`.  Just delete one `../` and it should work.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:17]:
```suggestion
    "In the real world the particle source for charged particle radiography experiments is finite in size and often has a non-uniform angular velocity distribution. For this reason, the [SyntheticProtronRadiography] particle tracing algorithm allows users to specify their own initial particle positions and velocities. This example will demonstrate how to use this functionality."
```

I still find this lead in a bit clunky.  Every time I get to the middle of the first sentence I have to re-scan the sentence to determine what is trying to be said.

Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:17]: I agree, how about this 

In real charged particle radiography experiments, the finite size of the particle source limits the resolution of the radiograph. Realistic sources may also generate a non-uniform angular distribution of particles, superimposing a large scale non-uniform "source profile" on the radiograph. For these reasons, the 
[SyntheticProtronRadiography] particle tracing algorithm allows users to specify their own initial particle positions and velocities. This example will demonstrate how to use this functionality to create a more realistic synthetic radiograph that includes these finite source and source profile effects.

**Comment by pheuer** [comment]
@StanczakDominik @rocco8773 

I think this is ready! All the links are now working. I created the file `test_mathematics.py` in the formulary test folder to write an explicit test for `rot_a_to_b()`.

I think a developer docs page on how to do the links in notebooks is a good idea, since they really improve the documentation but are annoying to create! The learning curve is steep since you have to commit and wait for CI to run to get feedback on whether you're doing it correctly. For the record, here are three examples that finally worked:

For a function
```
[rot_a_to_b()]: ../api/plasmapy.formulary.mathematics.rot_a_to_b.rst#plasmapy.formulary.mathematics.rot_a_to_b
```

For a class
```
[SyntheticProtronRadiography]: ../api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.rst#plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph
```

For a method of a class
```
[create_particles()]: 
../api/plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.rst#plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph.create_particles
```

For an external website
```
[np.random.choice()]: https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html
```

It would be worth explaining where the `api/` files are created. These paths are relative to the main `notebooks` directory: if the notebook is in a sub folder (eg. `cartesian_grids.ipynb` which is in `plasma`) an additional `../` is required to get back to the `api/` directory.

File: plasmapy/formulary/tests/test_mathematics.py
Comment (StanczakDominik) [inline plasmapy/formulary/tests/test_mathematics.py:19]:
```suggestion
    np.testing.assert_allclose(R, correct, atol=1e-6)
```
https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html has a bit nicer output when something breaks, so i like it for array comparisons

**Review by StanczakDominik — COMMENTED** [review]
Looked through non notebook changes; they look great. I'll leave the final call to @rocco8773, though (unless you're short on time, then I can continue here).

**Comment by rocco8773** [comment]
I got a partial review in the works, nothing too crazy. However, I ran out of steam tonight, so I'll post in the morning.

**Comment by StanczakDominik** [comment]
Sleep well!

**Review by rocco8773 — CHANGES_REQUESTED** [review]
This is a review of the custom source notebook.  All in all it looks good and I've confirmed the links are working.   All my suggestions focus on readability and clarity.  I just felt a few spots here and there could be refined a little more to make it easier for user to understand the notebook.

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:97]:
```suggestion
def rot_a_to_b(a: np.ndarray, b: np.ndarray) -> np.ndarray:
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:100]:
```suggestion
    Calculates the 3D rotation matrix that will rotate vector ``a`` to be aligned
    with vector ``b``.
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:105]:
```suggestion
    a : `~numpy.ndarray`, shape (3,)
        Vector to be rotated.  Should be a 1D, 3-element unit vector.  If ``a``
        is not normalize, then it will be normalized.
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:108]:
```suggestion
    b : `~numpy.ndarray`, shape (3,)
        Vector representing the desired orientation after rotation.  Should be
        a 1D, 3-element unit vector.  If ``b`` is not normalized, then it will
        be.
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:114]:
```suggestion
    R : `~numpy.ndarray`, shape (3,3)
        The rotation matrix that will rotate vector ``a`` onto vector ``b``.
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:117]:
```suggestion
    Notes
    -----
    The algorithm is based on
    `this discussion <https://math.stackexchange.com/a/476311>`_ on StackExchange:
```

Note, for linking in `sphinx` you need to have the trailing underscore `` `label <target>`_ ``.

I think you should also include the associated math in this "Notes" section.

File: plasmapy/formulary/tests/test_mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/tests/test_mathematics.py:40]:
```suggestion
```
This is not needed.

File: plasmapy/formulary/tests/test_mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/tests/test_mathematics.py:33]: In practice you want to write your tests so if one fails, then the other will still run.  In how you have the test written here, if the first test fails then the last two never run.  The best way to accomplish this is to use [`pytest.mark.parametrize`](https://docs.pytest.org/en/stable/parametrize.html), which would look a little like...

```python
@pytest.mark.parametrize(
    "a, b, correct",
    [
        # case 1
        (a, b, correct),
        # case 2
        (a, b, correct),
        ...
    ],
)
def test_rot_a_to_b(a, b, correct):
    R = rot_a_to_b(a, b)
    assert np.allclose(R, correct)
```

File: plasmapy/formulary/tests/test_mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/tests/test_mathematics.py:12]: When you add the shape conditioning for `a` and `b`, you'll also need to add a test for it.  I would typically to this like....

```python
@pytest.mark.parametrize(
    "a, b, _raises",
[
        # case 1
        (a, b, ValueError),
        # case 2
        (a, b, ValueError),
        ...
    ],
)
def test_rot_a_to_b_raises(a, b, _raises):
    with pytest.raises(_raises):
        rot_a_to_b(a, b)
```

**Review by rocco8773 — CHANGES_REQUESTED** [review]
This is a review of `rot_a_to_b` and its tests.

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:539]: Made a slight variation: 
```
Calling the [synthetic_radiograph()] function with the `ignore_grid` keyword will produce the synthetic radiograph corresponding to the source profile propagated freely through space (i.e. in the absence of any grid fields).
```

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:361]: Sure, done

File: docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb
Comment (pheuer) [reply docs/notebooks/proton_radiography_particle_tracing_custom_source.ipynb:465]: That's just one of the default printouts from `run()`: I agree it's not really relevant. How about I just set `verbose=False` and hide all of that?

**Comment by pheuer** [comment]
@rocco8773 Great, thanks for the review. I think I've addressed all your comments

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:129]:
```suggestion
    The algorithm is based on
    `this discussion <https://math.stackexchange.com/a/476311>`_ on StackExchange.
```
Use the shorter link so the docstrings stay within our max line width.

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:155]: It's not sufficient to rely on `np.dot` to condition `a` and `b` since it will still allow any N-element arrays to pass as long as `a.shape == b.shape`.  The functionality you've coded here only works for 1D, 3-element arrays so you need a workflow like...

```python
a = np.squeeze(a)  # will convert list to array and reduce any single element dimensions
if a.shape != (3,):
    raise ValueError("give message")
a = a / np.linalg.norm(a)
```

File: plasmapy/formulary/tests/test_mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/tests/test_mathematics.py:38]: Add a case where both are the wrong length.

File: plasmapy/formulary/mathematics.py
Comment (pheuer) [reply plasmapy/formulary/mathematics.py:156]: Ok, done

File: plasmapy/formulary/tests/test_mathematics.py
Comment (pheuer) [reply plasmapy/formulary/tests/test_mathematics.py:38]: Done

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:155]:
```suggestion
        raise ValueError(
            f"Argument 'a' must have shape (3,) but input has shape {a.shape}."
        )
```

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:160]:
```suggestion
        raise ValueError(
            f"Argument 'b' must have shape (3,) but input has shape {b.shape}."
        )
```

File: plasmapy/formulary/tests/test_mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/tests/test_mathematics.py:40]: Do both are wrong length but equal length.  This is the case that should have failed, but would have passed if you were just depending on `np.dot`.

File: plasmapy/formulary/mathematics.py
Comment (rocco8773) [inline plasmapy/formulary/mathematics.py:132]:
```suggestion
```

Delete extra space.  Sorry for throwing all this little things at you.  Just trying to catch it now so we don't have to come back.

File: plasmapy/formulary/tests/test_mathematics.py
Comment (pheuer) [reply plasmapy/formulary/tests/test_mathematics.py]: Got it- done

File: docs/notebooks/proton_radiography_particle_tracing.ipynb
Comment (rocco8773) [inline docs/notebooks/proton_radiography_particle_tracing.ipynb:262]:
```suggestion
```

This link is never used so you can delete it.

File: changelog/1054.doc.rst
Comment (rocco8773) [inline changelog/1054.doc.rst:1]:
```suggestion
Created an example notebook demonstrating how the
`~plasmapy.diagnostics.proton_radiography.SyntheticProtonRadiograph` class can be
used to generate synthetic proton radiographs with arbitrary source profiles.  Add
code documentation links to all proton radiograph notebooks.
```

This will allow `intersphinx` to properly generate the link for `SyntheticProtonRadioraph`.

**Review by rocco8773 — CHANGES_REQUESTED** [review]
Thanks for sticking with this.  Everything looks really good!

That last thing that's needed is a change log entry for `rot_a_to_b`, so create a `1054.feature.rst` for this.

**Comment by pheuer** [comment]
No problem - change log entry created!

**Review by rocco8773 — APPROVED** [review]
All looks good.  I'm going to wait for the checks to pass and then I'll merge. Thanks for the hard work!

**Pull request merged**
